src/python/pants/backend/scala/lint/scalafix/rules.py

```diff
@@ -0,0 +1,382 @@
+from pants.backend.scala.subsystems.scalac import Scalac
+from pants.backend.scala.target_types import ScalaSourceField
+from pants.backend.scala.util_rules.versions import ScalaVersion
+from pants.core.goals.fix import FixResult, FixTargetsRequest, Partitions
+from pants.core.goals.generate_lockfiles import GenerateToolLockfileSentinel
+from pants.core.goals.lint import LintResult, LintTargetsRequest
+from pants.core.util_rules.config_files import (
+    GatherConfigFilesByDirectoriesRequest,
+    GatheredConfigFilesByDirectories,
+)
+from pants.core.util_rules.partitions import Partition
+from pants.core.util_rules.stripped_source_files import _stripped_snapshot_by_source_roots
+from pants.engine.addresses import Addresses, UnparsedAddressInputs
+from pants.engine.fs import AddPrefix, Digest, DigestSubset, MergeDigests, PathGlobs, Snapshot
+from pants.engine.process import FallibleProcessResult
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.target import FieldSet, Target
+from pants.engine.unions import UnionRule
+from pants.jvm.classpath import Classpath
+from pants.jvm.compile import ClasspathEntry
+from pants.jvm.goals import lockfile
+from pants.jvm.jdk_rules import InternalJdk, JvmProcess
+from pants.jvm.resolve.coursier_fetch import ToolClasspath, ToolClasspathRequest
+from pants.jvm.resolve.jvm_tool import GenerateJvmLockfileFromTool, GenerateJvmToolLockfileSentinel
+from pants.jvm.resolve.key import CoursierResolveKey
+from pants.source.source_root import SourceRootsRequest, SourceRootsResult
+from pants.util.dirutil import group_by_dir
+from pants.util.frozendict import FrozenDict
+from pants.util.logging import LogLevel
+from pants.util.strutil import Simplifier, pluralize
+
+
+@dataclass(frozen=True)
+class ScalafixFieldSet(FieldSet):
+    required_fields = (ScalaSourceField,)
+
+    source: ScalaSourceField
+
+    @classmethod
+    def opt_out(cls, tgt: Target) -> bool:
+        return tgt.get(SkipScalafixField).value
+
+
+class ScalafixFixRequest(FixTargetsRequest):
+    field_set_type = ScalafixFieldSet
+    tool_subsystem = ScalafixSubsystem
+
+
+class ScalafixLintRequest(LintTargetsRequest):
+    field_set_type = ScalafixFieldSet
+    tool_subsystem = ScalafixSubsystem
+
+
+@dataclass(frozen=True)
+class ScalafixPartitionInfo:
+    scala_version: ScalaVersion | None
+    config_snapshot: Snapshot
+    runtime_classpath_entries: tuple[str, ...]
+    compile_classpath_entries: tuple[str, ...]
+    rule_classpath_entries: tuple[str, ...]
+    extra_immutable_input_digests: FrozenDict[str, Digest]
+
+    @property
+    def description(self) -> str:
+        return self.config_snapshot.files[0]
+
+
+class ScalafixToolLockfileSentinel(GenerateJvmToolLockfileSentinel):
+    resolve_name = ScalafixSubsystem.options_scope
+
+
+@dataclass(frozen=True)
+class _ScalafixRuleClasspath:
+    classpath: Classpath | None
+
+    def args(self, *, prefix: str = "") -> Iterator[str]:
```

Comment:
Why not return `Iterable` instead and then you would not need to do `iter([])`, just `[]`?